lib/sycamore/sycamore/transforms/summarize.py

```diff
@@ -143,7 +141,238 @@ def collapse(text: str, tokens_per_chunk: int, tokenizer: Tokenizer, summarizer_
+        self.element_batch_size = element_batch_size
+
+    def get_const_vars(self) -> dict[str, str]:
+        return {
+            "iteration_var": "_summarize_round",
+            "num_elements_key": "_num_total_elements",
+            "index_key": "_summarizer_element_index",
+            "intermediate_summary_key": "_partial_summary",
+        }
+
+    def as_llm_map(self, child: Optional[Node], **kwargs) -> Node:
+        vars = self.get_const_vars()
+
+        prompt = self.prompt.set(
+            iteration_var=vars["iteration_var"],
+            intermediate_summary_key=vars["intermediate_summary_key"],
+            index_key=vars["index_key"],
+        )
+        if self.element_batch_size is not None:
+            prompt = prompt.set(branching_factor=self.element_batch_size)
+        if self.fields is not None:
+            prompt = prompt.set(fields=self.fields)
+        if self.question is not None:
+            prompt = prompt.set(question=self.question)
+
+        def preprocess(doc: Document) -> Document:
+            for i, e in enumerate(doc.elements):
+                e.properties[vars["num_elements_key"]] = len(doc.elements)
+                e.properties[vars["index_key"]] = i
+            return doc
+
+        def validate(elt: Element) -> bool:
+            iv = elt.properties[vars["iteration_var"]]
+            num_elts = elt.properties[vars["num_elements_key"]]
+            assert isinstance(prompt, JinjaElementPrompt)
+            branching_factor = prompt.kwargs["branching_factor"]
+            return branching_factor ** (iv + 1) > num_elts
+
+        def cleanup(doc: Document) -> Document:
+            if len(doc.elements) == 0:
+                doc.properties["summary"] = ""
+                return doc
+            found_summary = False
+            for e in doc.elements:
+                if not found_summary and vars["intermediate_summary_key"] in e.properties:
+                    doc.properties["summary"] = doc.elements[0].properties[vars["intermediate_summary_key"]]
+                    found_summary = True
+                for k in vars:
+                    if k in e.properties:
+                        del e.properties[k]
+            if not found_summary:
+                doc.properties["summary"] = ""
+            return doc
+
+        premap = Map(child, f=preprocess)
+        llm_map = LLMMapElements(
+            child=premap,
+            prompt=prompt,
+            output_field=vars["intermediate_summary_key"],
+            llm=self.llm,
+            iteration_var=vars["iteration_var"],
+            validate=validate,
+            max_tries=20,  # If you hit this then your document has at least 2^20 elements in it.
+        )
+        postmap = Map(llm_map, f=cleanup)
+        comptransform = CompositeTransform(child, [])  # type: ignore
+        comptransform.nodes = [premap, llm_map, postmap]
+        return comptransform
+
+
+class MaxTokensHeirarchicalDocumentSummarizer(Summarizer):
```

Comment:
Big document, spread evenly across context windows of packed with a tail?
Multiple documents/ split only at document boundary or packed.
Split at properties or not?
How to split properties if they exceed context window.
Spread documents evenly or not, e.g. with 10 docs => 5,4,1; or 4,3,3?